Add LimitedOffsetsEnum to Limited offset token

luyuncheng opened the pull request:
In #67325 add `max_analyzed_offset` allows users to limit the highlighting of text fields.
but when using `"index_options": "offsets"` in mappings, this `max_analyzed_offset offset` can not be limited, 
because it would use `OffsetSource.ANALYSIS`and `PostingsOffsetStrategy` to highlight field.

When a string="Testing Fun Testing Fun" need to be highlight

Using `UnifiedHighlighter.OffsetSource.ANALYSIS` with `queryMaxAnalyzedOffset=10` would response: 
```
"Testing <b>Fun</b> Testing Fun"
```
because in #67325 add `LimitTokenOffsetAnalyzer`


__BUT USING__ `UnifiedHighlighter.OffsetSource.POSTINGS` which mapping is `index_options=offsets` and `queryMaxAnalyzedOffset=10` would response:
```
"Testing <b>Fun</b> Testing <b>Fun</b>"
```
because offset can not be limited by `LimitTokenOffsetAnalyzer`

may be can add a new `LimitedOffsetsEnum` to limit offset token

Link Issue: #86109

luyuncheng committed:
1. Add LimitedOffsetsEnum to Limited offset token
2. Add highlight tests for maxAnalyzedOffset in different OffsetSource

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

rahuldimri commented:
is this issue available to work ?. or is there anything pending on this issue?

romseygeek commented on server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java:
```diff
@@ -394,4 +396,102 @@ public void testExceedMaxAnalyzedOffset() throws Exception {
+
+
+    protected Analyzer wrapAnalyzer(Analyzer analyzer, Integer maxAnalyzedOffset) {
+        if (maxAnalyzedOffset != null) {
```
Why do you need to wrap the analyzer here? Shouldn't it be handled by passing the max offset to CustomUnifiedHighlighter?

luyuncheng replied:
At beginning i want to describe how the analyzer runs with different analyzers(Default VS LimitTokenOffsetAnalyzer) works in different maxAnalyzedOffset. 
i think in this test case is unnecessary, but we can add random tests for different  maxAnalyzedOffset

romseygeek replied:
++, random tests would be great

romseygeek commented:
@elasticmachine test this please

luyuncheng committed:
Merge branch 'elastic:main' into LIMIT_CustomUnifiedHighlighter_TOKENS

luyuncheng committed:
1. Add changelog AND Random tests for different offset source in CustomUnifiedHighlighterTests

luyuncheng commented:
> ++, random tests would be great


@romseygeek  Thanks for the reviewing this.

At https://github.com/elastic/elasticsearch/pull/86110/commits/5d5e7018cb85dfee39f4a94844186f6319bf3e86 i removed the `wrapAnalyzer` and merge the assertHighlightOneDoc into one function
AND
i added a test for random different `OffsetSource` and `maxAnalyzedOffset `

romseygeek commented on server/src/test/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighterTests.java:
```diff
@@ -394,4 +425,73 @@ public void testExceedMaxAnalyzedOffset() throws Exception {
+            Locale.ROOT,
+            BreakIterator.getSentenceInstance(Locale.ROOT),
+            0,
+            // OLD Strategy would Response： "Testing <b>Fun</b> Testing <b>Fun</b>"
```
This comment will be confusing in a couple of years time (what was the old strategy? why was it different?).  I think the current behaviour is a bug so it's fine to just test for it, no need to add a comment.

luyuncheng replied:
LGTM

romseygeek reviewed (changes requested):
Thanks @luyuncheng! I left a few small comments but I think this is close.

luyuncheng committed:
Fixed CustomFieldHighlighter for param AND remove some comment

luyuncheng committed:
Remove unnecessary import

luyuncheng commented on server/src/main/java/org/elasticsearch/lucene/search/uhighlight/LimitedOffsetsEnum.java:
```diff
@@ -0,0 +1,57 @@
+    private final OffsetsEnum delegate;
+    private final Integer maxOffset;
+
+    public LimitedOffsetsEnum(OffsetsEnum delegate, @Nullable Integer maxOffset) {
```
LGTM

luyuncheng replied:
In https://github.com/elastic/elasticsearch/pull/86110/commits/e18513f96d5edf5887329cf3b73927dd0587820a fixed it

romseygeek commented:
@elasticmachine test this please

romseygeek commented:
@elasticmachine generate changelog

romseygeek commented on docs/changelog/86110.yaml:
```diff
@@ -0,0 +1,6 @@
+pr: 86110
+summary: Add LimitedOffsetsEnum to Limited offset token
+area: Search/Highlighting
```
Can you change this to just `Search`? Apparently `Search/Highlighting` doesn't work here, 🤷

luyuncheng replied:
OK

luyuncheng committed:
Update 86110.yaml

Fixed chagelog area

luyuncheng commented:
@elasticmachine  test this please

luyuncheng committed:
Merge branch 'elastic:main' into LIMIT_CustomUnifiedHighlighter_TOKENS

romseygeek commented:
@elasticmachine ok to test

romseygeek commented:
Hi @luyuncheng can you run `./gradlew precommit` at the root and then push the relevant changes?

luyuncheng committed:
Merge pull request #2 from luyuncheng/master

Merge Master into it

luyuncheng committed:
precommit check

luyuncheng committed:
Merge pull request #3 from luyuncheng/LIMIT_CustomUnifiedHighlighter_TOKENS_PRE

Merge Master and precommit into it

romseygeek commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into LIMIT_CustomUnifiedHighlighter_TOKENS

romseygeek commented:
Thanks for your patience @luyuncheng!
